fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,576 @@
+        newOutputProperty.setDistributionProperty(newDistributionProperty);
+        GroupExpression enforcer = newDistributionProperty.appendEnforcers(childGroup);
+
+        enforcer.setOutputPropertySatisfyRequiredProperty(newOutputProperty, newOutputProperty);
+        updateChildCostWithEnforcer(enforcer, oldOutputProperty, newOutputProperty, childCost, childGroup);
+        return newOutputProperty;
+    }
+
+    private void updateChildCostWithEnforcer(GroupExpression enforcer,
+                                             PhysicalPropertySet oldOutputProperty,
+                                             PhysicalPropertySet newOutputProperty,
+                                             double childCost, Group childGroup) {
+        context.getMemo().insertEnforceExpression(enforcer, childGroup);
+        // update current total cost
+        curTotalCost -= childCost;
+        // add enforcer cost
+        childCost += CostModel.calculateCost(enforcer);
+        curTotalCost += childCost;
+
+        enforcer.setPropertyWithCost(newOutputProperty, Lists.newArrayList(oldOutputProperty), childCost);
+        childGroup.setBestExpression(enforcer, childCost, newOutputProperty);
+
+        if (ConnectContext.get().getSessionVariable().isSetUseNthExecPlan()) {
+            enforcer.addValidOutputInputProperties(newOutputProperty, Lists.newArrayList(PhysicalPropertySet.EMPTY));
+            enforcer.getGroup().addSatisfyRequiredPropertyGroupExpression(newOutputProperty, enforcer);
+        }
+    }
+
+    private void transToBucketShuffleJoin(HashDistributionDesc leftLocalDistributionDesc,
+                                          List<Integer> leftShuffleColumns, List<Integer> rightShuffleColumns) {
+        List<Integer> bucketShuffleColumns = Lists.newArrayList();
+        for (int leftScanColumn : leftLocalDistributionDesc.getColumns()) {
+            int index = leftShuffleColumns.indexOf(leftScanColumn);
+            bucketShuffleColumns.add(rightShuffleColumns.get(index));
+        }
+
+        DistributionSpec rightDistributionSpec =
+                DistributionSpec.createHashDistributionSpec(new HashDistributionDesc(bucketShuffleColumns,
+                        HashDistributionDesc.SourceType.BUCKET_JOIN));
+
+        GroupExpression rightChild = childrenBestExprList.get(1);
+        PhysicalPropertySet rightChildOutputProperty = childrenOutputProperties.get(1);
+        // enforce right child BUCKET_JOIN type distribution
+        enforceChildDistribution(rightDistributionSpec, rightChild, rightChildOutputProperty);
+
+        // update group expression require property
+        PhysicalPropertySet newRightChildInputProperty = createPropertySetByDistribution(rightDistributionSpec);
+        requiredChildrenProperties.set(1, newRightChildInputProperty);
+    }
+
+    // enforce child SHUFFLE type distribution
+    private void enforceChildShuffleDistribution(List<Integer> shuffleColumns, GroupExpression child,
```

Comment:
Why we have enforce operator in property deriver？

Reply:
compute the hash join node output property, StarRocks support the colocate shuffle join, this join type  not only do child nodes need to satisfy a particular data distribution, but some additional checks are required(see canColocateJoin). If can not satisfy, It needs to add enforcer to make sure it's an legitimate  plan. Hint is similar, with an extra enforcer